src/compiler/rust.rs

```diff
@@ -300,19 +300,41 @@ where
     Ok(parse_dep_info(&deps, cwd))
 }
 
-fn parse_dep_info<T>(dep_info: &str, cwd: T) -> Vec<PathBuf>
+fn parse_dep_info<T>(dep_info: &str, cwd: T) -> (Vec<PathBuf>, Vec<(OsString, OsString)>)
 where
     T: AsRef<Path>,
 {
     let cwd = cwd.as_ref();
     // Just parse the first line, which should have the dep-info file and all
     // source files.
-    let line = match dep_info.lines().next() {
-        None => return vec![],
+    let mut lines = dep_info.lines();
+    let line = match lines.next() {
+        None => return (vec![], vec![]),
         Some(l) => l,
     };
+
+    // collect the env variables referenced via `env!` or required by `build.rs`.
+    const ENV_DEP_PREFIX: &str = "# env-dep:";
+    let mut dep_envs = lines
+        .filter(|line| line.starts_with(ENV_DEP_PREFIX) && line.len() > ENV_DEP_PREFIX.len())
```

Comment:
It'd be good to use `strip_prefix` API but since that's introduced in 1.45 it'd be great to leave a note stating that we should upgrade to that once we bump MSRV